### PR TITLE
[Preview] Pyodide shimming of requests

### PR DESCRIPTION
**Pull request by bartbroere**
Update: Binary responses have been fixed. Moving to `fetch` hasn't been done yet.

~This requests shim functions for basic GET and POST request, but it doesn't work for binary or streaming responses. The main reason for this is that `XMLHttpRequest` is used, instead of the more modern `fetch` API in Javascript. On top of that, it runs in synchronous mode, mainly because the `requests` module isn't an async library.~

As some kind of roadmap I would like to move to `fetch`, while keeping the `requests` API intact. Ideas on how to achieve calling the asynchronous `fetch` in Javascript from the synchronous shimmed methods in Python `requests` are very welcome!

~By moving to `fetch` instead of `XMLHttpRequest`, we automatically would create the ability to receive binary / streaming responses (`stream=True`).~

### CODE REVIEW

File: requests/sessions.py [hunk 1]
```diff
@@ -512,36 +430,40 @@ def request(self, method, url,
...
-            'allow_redirects': allow_redirects,
-        }
-        send_kwargs.update(settings)
-        resp = self.send(prep, **send_kwargs)
-
-        return resp
+        request = XMLHttpRequest.new()
+        request.responseIsBinary = False
+        if stream:
+            request.overrideMimeType('text/plain; charset=x-user-defined')
```
Comment (bartbroere) [inline requests/sessions.py]:
```suggestion
            # we ask the browser not to worry about the character set, keeping the raw bytes intact
            request.overrideMimeType('text/plain; charset=x-user-defined')
```

File: requests/models.py [hunk 2]
```diff
@@ -603,24 +598,20 @@ class Response(object):
         'encoding', 'reason', 'cookies', 'elapsed', 'request'
     ]
 
-    def __init__(self):
+    def __init__(self, request):
+        if request.responseIsBinary:
+            self.raw = bytes(ord(byte) & 0xff for byte in request.response)
```
Comment (bartbroere) [inline requests/models.py]:
```suggestion
            # bring everything outside the range of a single byte within this range
            self.raw = bytes(ord(byte) & 0xff for byte in request.response)
```

**Comment by psychemedia** [comment]
Is this PR a more robust pyodide compatible version of `requests` than your `pyodide-requests` package?

**Comment by bartbroere** [comment]
> Is this PR a more robust pyodide compatible version of `requests` than your `pyodide-requests` package?

Yes, good question! This is indeed the version I'll continue to work on. There's still quite some issues to tackle, but at least this is the maintained version.

To make it a bit clearer, I've archived the `pyodide-requests` repository.

**Comment by prescod** [comment]
How do you actually add this to Node or Web projects?

Do you publish it on pypi?

**Comment by bartbroere** [comment]
> How do you actually add this to Node or Web projects?
> 
> 
> 
> Do you publish it on pypi?

You would use this in Node or web projects combined with the Pyodide library. Since some time the `requests` library is included in Pyodide, so this patch is no longer needed.

**Pull request closed without merging**